src/java/org/apache/cassandra/db/compaction/unified/Controller.java

```diff
@@ -202,36 +249,180 @@ public int getThreshold(int index) {
+            if (logger.isDebugEnabled())
+                logger.debug("Shard count {} for density {}, {} times target {}",
+                             shards,
+                             FBUtilities.prettyPrintBinary(localDensity, "B", " "),
+                             localDensity / targetSSTableSize,
+                             FBUtilities.prettyPrintBinary(targetSSTableSize, "B", " "));
+            return shards;
+        }
+        else
+        {
+            // How many we would have to aim for the target size. Divided by the base shard count, so that we can ensure
+            // the result is a multiple of it by multiplying back below.
+            double count = localDensity / (targetSSTableSize * baseShardCount);
+            // Take a logarithm of the count (in base 2), and adjust it by the given growth modifier.
+            // Adjust by 0.5 to round the exponent so that the result falls between targetSSTableSize * sqrt(0.5) and
+            // targetSSTableSize * sqrt(2). Finally, make sure the exponent is at least 0 and not greater than the
+            // fixed maximum.
+            // Note: This code also works correctly for the special cases of sstableGrowthModifier == 0 and 1,
+            //       but the above code avoids the floating point arithmetic for these common cases.
+            // Note: We use log instead of getExponent because we also need the non-integer part of the logarithm
+            //       in order to apply the growth modifier correctly.
+            final double countLog = Math.log(count);
+            double pow = countLog * INVERSE_LOG_2 * (1 - sstableGrowthModifier) + 0.5;
+            if (pow >= MAX_SHARD_SHIFT)
+                shards = baseShardCount << MAX_SHARD_SHIFT;
+            else if (pow >= 0)
+                shards = baseShardCount << (int) pow;
+            else
+                shards = baseShardCount;    // this also covers the case of pow == NaN
+
+            if (logger.isDebugEnabled())
+            {
+                long targetSize = (long) (targetSSTableSize * Math.exp(countLog * sstableGrowthModifier));
+                logger.debug("Shard count {} for density {}, {} times target {}",
+                             shards,
+                             FBUtilities.prettyPrintBinary(localDensity, "B", " "),
+                             localDensity / targetSize,
+                             FBUtilities.prettyPrintBinary(targetSize, "B", " "));
+            }
+            return shards;
+        }
+    }
+
+    private boolean isPowerOf2(int n)
+    {
+        return (int)(Math.ceil((Math.log(n) / Math.log(2)))) == (int)(Math.floor(((Math.log(n) / Math.log(2)))));
+    }
+
+    /**
+     * @param n number of shards needs to be divisible by n
+     * @param shards the maximum number of shards
+     * @return the largest power of 2 that is a divisor of n (e.g. when n = 10, return 2) and less than shards
+     */
+    private int getLargestPowerOf2Divisor(int n, int shards)
```

Comment:
The largest power-of-2 divisor of `n` can be found via `Integer.numberOfTrailingZeros(n)`. 
I.e. this can be implemented as `Math.min(shards, 1 << Integer.numberOfTrailingZeros(n))`.

This would also work without an explicit `isPowerOf2` check.